### PR TITLE
feat(batch): parallel table scan

### PR DESCRIPTION
**Pull request by xxchan**
## What's changed and what's your intention?

Mainly added vnode_bitmap to ExecutorBuilder.

Not finished yet... See the `todo!`s
- schedule: how to partition vnode bitmap for different executors?
- executor: scan different ranges according to vnode bitmap
 
BTW, should we use vnode bitmap or other representations? e.g., vnode ranges, maybe like `ParallelUnitMapping`. I used bitmap here because of previous storage implementation 😇

## Checklist

- [x] I have written necessary docs and comments
- [x] I have added necessary unit tests and integration tests
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Refer to a related PR or issue link (optional)
#3237

### CODE REVIEW

**Comment by xxchan** [comment]
Just tried another representation [use vnode_ranges instead of vnode_bitmap](https://github.com/singularity-data/risingwave/pull/3251/commits/9255aa615ec4b6fde0ec754a9ed8c73bc460f34d), which may be easier for executors to create vnode prefix range iterators

**Comment by xxchan** [comment]
Just tried another representation [use vnode_ranges instead of vnode_bitmap](https://github.com/singularity-data/risingwave/pull/3251/commits/9255aa615ec4b6fde0ec754a9ed8c73bc460f34d), which may be easier for executors to create vnode prefix range iterators

File: proto/task_service.proto
Comment (liurenjie1024) [inline proto/task_service.proto:38]: Why `vnode_ranges` is an attribute of task? I think it's the attribute of table scan plan node.

File: proto/task_service.proto
Comment (liurenjie1024) [inline proto/task_service.proto:84]: Why the response need to know `vnode_ranges`?

File: src/batch/src/execution/grpc_exchange.rs
Comment (liurenjie1024) [inline src/batch/src/execution/grpc_exchange.rs:40]: The `vnode_range` should be determined by optimier/fragmenter.

File: proto/task_service.proto
Comment (liurenjie1024) [reply proto/task_service.proto:38]: BTW, would it be more general to make it a `VNode` list?

File: src/batch/src/execution/grpc_exchange.rs
Comment (xxchan) [reply src/batch/src/execution/grpc_exchange.rs:40]: I think it should be determined by scheduler? Each scan task scans different ranges

File: proto/task_service.proto
Comment (xxchan) [reply proto/task_service.proto:38]: > Why vnode_ranges is an attribute of task? I think it's the attribute of table scan plan node.

Because it's the result of scheduling - Each scan task scans different ranges. Putting it in scan plan node is also possible, but requires rewriting each task's plan node. Both ok to me.

> BTW, would it be more general to make it a VNode list?

Previously it's vnode bitmap, but I think list/bitmap may be difficult for scan executor to use?

File: src/batch/src/execution/grpc_exchange.rs
Comment (liurenjie1024) [reply src/batch/src/execution/grpc_exchange.rs:40]: I think vnodes of a table scan plan node should be determined by optimizer. And the compute node where a task  should be sent to determined fragmenter, this way we can reuse this logic in local execution mode. The scheduler should only care about sending task to right compute node.

Comment (xxchan) [reply src/batch/src/execution/grpc_exchange.rs:40]: For optimizer & fragmenter, they only have one plan. e.g., scan full table

Then the scan is divided into non-overlapping vnode ranges, and each task of the fragment is assigned a range to scan. How can we do that in optimizer/fragmenter?

Comment (xxchan) [reply src/batch/src/execution/grpc_exchange.rs:40]: Oh we can let fragmenter store `num_parallelsm` different plans...

Comment (xxchan) [reply src/batch/src/execution/grpc_exchange.rs:40]: > And the compute node where a task should be sent to determined fragmenter

Isn't this called "scheduling"? ... 

> The scheduler should only care about sending task to right compute node. 

If so, maybe it should not be called "scheduler"

Comment (fuyufjh) [reply src/batch/src/execution/grpc_exchange.rs:40]: > I think it should be determined by scheduler? Each scan task scans different ranges

Agree with xxchan. The planner should never care about the specifical data distribution (like, what key is in which parallel unit). It should be done on execution.

Comment (liurenjie1024) [reply src/batch/src/execution/grpc_exchange.rs:40]: 1. For optimizer, it's supposed to figure out vnodes for each table scan. Think about this: 
```
select * from t where key in (1, 2);
```
Here after the filter has been pushed to table scan, it should be able to prune unnecessary vnodes.
2. Yes, fragmenter should be able to figure out number of parallelism of each stage, and assign prefer location to each partition(e.g., vnode range). For this case , we should assign a list of partitions to leaf stage with table scan, where each partition has two fields:
```
vnode list
location
```
3.  The scheduler should schedule tasks according to parallelism and partitions.

Comment (fuyufjh) [reply src/batch/src/execution/grpc_exchange.rs:40]: > Here after the filter has been pushed to table scan, it should be able to prune unnecessary vnodes.

Optimizer is responsible for pushing predicate (SARGS) to table scan, but is not responsible for deciding the specific vnode number.

File: proto/task_service.proto
Comment (fuyufjh) [inline proto/task_service.proto:38]: Should it be a list? like 

```
repeated common.VNodeRanges vnode_ranges = 4;
```

File: src/batch/src/execution/grpc_exchange.rs
Comment (xxchan) [reply src/batch/src/execution/grpc_exchange.rs:40]: For 1, it's not the target of this PR and the `vnodes_range` introduced here. 

This PR cares mainly about *given a range* (e.g., full table scan), divide (schedule) it into small ranges and assign them to tasks. 

I think the `ScanRange` introduced earlier (in `BatchSeqScan`) is responsible for this part? And pruning vnodes can be done at the same place where we partition the vnode ranges, instead of optimizer

Comment (liurenjie1024) [reply src/batch/src/execution/grpc_exchange.rs:40]: Per my understanding, the parallel unit is just a partition, which should be maintained by optimizer. It should not know vnode, but should understand parallel unit.

Comment (fuyufjh) [reply src/batch/src/execution/grpc_exchange.rs:40]: > Per my understanding, the parallel unit is just a partition, which should be maintained by optimizer. It should not know vnode, but should understand parallel unit.

I don't think it as partition 😇 To me, it's exactly same thing to the metadata of KV databases like Cassendra, HBase or TikV

File: proto/task_service.proto
Comment (xxchan) [reply proto/task_service.proto:38]: `VNodeRanges` is multiple ranges
```
// Records the vnodes owned by a parallel unit.
//
// `starts`: [s1, s2, ...], `ends`: [e1, e2, ...] means ranges [s1, e1], [s2, e2], ... (both ends included)
message VNodeRanges {
  repeated uint64 starts = 1;
  repeated uint64 ends = 2;
}
```

It can be sth like `repeated VNodeRange` or `VNodeRanges`. I just drafted it without too much consideration 🥸

File: src/common/src/consistent_hashing.rs
Comment (xx01cyx) [inline src/common/src/consistent_hashing.rs:32]: We may not use vnode mappings built here. Instead, we have to use vnode mappings built in meta when an mv is created. I'll implement this (inform frontend of the vnode mappings) soon.

File: src/batch/src/execution/grpc_exchange.rs
Comment (liurenjie1024) [reply src/batch/src/execution/grpc_exchange.rs:40]: Anyway, we need to map `filters in table scan` to `parallel units`, and it can't be don't in scheduler since we need to reuse it in local mode. Maybe fragmenter is a better place.

File: proto/task_service.proto
Comment (BugenZhao) [reply proto/task_service.proto:38]: We're introducing yet another representation! 🤣

Comment (BugenZhao) [reply proto/task_service.proto:38]: +1 for `repeated Range`, then `zip_`**`eq`** is implied.

File: src/batch/src/task/task_manager.rs
Comment (BugenZhao) [inline src/batch/src/task/task_manager.rs:248]: Seems we should use full range instead of an empty one as a fallback?

File: src/frontend/src/scheduler/distributed/stage.rs
Comment (BugenZhao) [inline src/frontend/src/scheduler/distributed/stage.rs:296]: Will it pass the clippy check as we introduce `todo!` in the middle?

File: src/common/src/catalog/physical_table.rs
Comment (xx01cyx) [inline src/common/src/catalog/physical_table.rs:42]:
```suggestion
    pub vnode_mapping: Option<Vec<ParallelUnitId>>,
```
My fault... Forgot to replace `u32` with `ParallelUnitId` in last PR...

**Comment by xxchan** [comment]
PTAL the partitioning logic in distributed scheduler (`stage.rs` `schedule_tasks`) & local mode (`local.rs` `convert_plan_node`) first. I'm not sure whether I understand it correctly 😇 (Whether the representation looks good/can be placed somewhere else can be further discussed if there's no other big problems)


And the remaining work is the executor part, which is blocked by storage side's work (new encoding & vnode ranges scan API).

File: src/frontend/src/scheduler/distributed/stage.rs
Comment (BowenXiao1999) [inline src/frontend/src/scheduler/distributed/stage.rs:296]: What about add some comments here why we need a if-else branch here?

**Comment by BowenXiao1999** [comment]
> how to partition vnode bitmap for different executors?

I think you mean different tasks? cuz there is only tabel scan need vnode bitmap?

**Comment by xxchan** [comment]
> > how to partition vnode bitmap for different executors?
> 
> 
> 
> I think you mean different tasks? cuz there is only tabel scan need vnode bitmap?

Yes, I mean parallel tasks/executors for a scan plan node

File: proto/task_service.proto
Comment (liurenjie1024) [inline proto/task_service.proto:90]: Why the reponse need `vnode_ranges`?

File: proto/task_service.proto
Comment (liurenjie1024) [inline proto/task_service.proto:41]: I'm still not convinced why `vnode_ranges` should be task rather than plan node?

File: src/batch/src/executor/row_seq_scan.rs
Comment (liurenjie1024) [inline src/batch/src/executor/row_seq_scan.rs:160]: `vnode_ranges` not used?

File: src/frontend/src/scheduler/local.rs
Comment (liurenjie1024) [inline src/frontend/src/scheduler/local.rs:85]: This builder is root task executed in frontend, so we should not have any vnodes

**Review by liurenjie1024 — CHANGES_REQUESTED** [review]
I don't think we should put vnode ranges in task, but we should put in `TableScan` plan node.  For example one day we want to implement join push down, e.g. one task may have two table scans, we can't identify the correct vnodes.

File: src/batch/src/rpc/service/task_service.rs
Comment (liurenjie1024) [inline src/batch/src/rpc/service/task_service.rs:58]: Nice catch!

File: src/frontend/src/scheduler/plan_fragmenter.rs
Comment (liurenjie1024) [inline src/frontend/src/scheduler/plan_fragmenter.rs:389]: It's possible that one stage has  more than one table scans, for example:
```
select * from t1, t2 where t1.a = t2.a
```
When t1 and t2 are both distributed by `a` and colocated in same node, we have following plan:
```
                          Exchange
                                 |
                           HashJoin
                   /                           \
              TableScan(t1)       TableScan(t2)
                
```

**Comment by BugenZhao** [comment]
> The overall structure is ok to me. But I think there are some missing part:
> 
> 1. Create iterator using vnode ranges

I'm working on this. #3316

File: src/frontend/src/scheduler/distributed/stage.rs
Comment (xxchan) [reply src/frontend/src/scheduler/distributed/stage.rs:296]: added

**Comment by xxchan** [comment]
It runs!

File: src/frontend/src/scheduler/plan_fragmenter.rs
Comment (xxchan) [reply src/frontend/src/scheduler/plan_fragmenter.rs]: Currently it's not possible, since HashJoin has Exchange on both its inputs. What about doing it later when needed?

File: src/storage/src/table/cell_based_table.rs
Comment (BugenZhao) [inline src/storage/src/table/cell_based_table.rs:304]: I'd like to get #3558 merged first.

**Review by liurenjie1024 — APPROVED** [review]
Generally LGTM

File: src/frontend/src/scheduler/plan_fragmenter.rs
Comment (liurenjie1024) [reply src/frontend/src/scheduler/plan_fragmenter.rs]: Hmm, the PR is too large, so I'm ok to put it later.

**Review by BugenZhao — CHANGES_REQUESTED** [review]

File: src/storage/src/table/cell_based_table.rs
Comment (xxchan) [reply src/storage/src/table/cell_based_table.rs:283]: I mean https://github.com/singularity-data/risingwave/issues/3450 is not implemented, i.e., the executor will still be created even if its partition doesn't need to be read

Comment (xxchan) [reply src/storage/src/table/cell_based_table.rs:283]: But I can let the executor early exits now 🤔🥸

Comment (BugenZhao) [reply src/storage/src/table/cell_based_table.rs:283]: Do you mean point getting will still schedule multiple seq scan executors? I think the scheduler should only create one executor and assign full vnodes for it, instead of hacking on the cell-based table implementation.

Comment (BugenZhao) [reply src/storage/src/table/cell_based_table.rs:283]: Removing this assertion will seriously affect the debugging of distribution keys like #3573 #3572...

Comment (xxchan) [reply src/storage/src/table/cell_based_table.rs:283]: I was planning to do it in a next PR. But I can workaround it in executor now.

**Comment by xxchan** [comment]
e2e tests break due to ordering 😅

**Comment by BugenZhao** [comment]
> e2e tests break due to ordering 😅

Ordering from every parallel unit is guaranteed, while there seems to be no merge-sort among chunks from different parallel units. 🤤

**Comment by xxchan** [comment]
@BugenZhao PTAL the last commit

**Comment by xxchan** [comment]
forget to revert 😅

**Comment by xxchan** [comment]
PTAL the final updates:
- ordering: https://github.com/singularity-data/risingwave/pull/3582 , #3583
  - **FIXME**: parallel scan on ordered mv may have unorderd results. should do a merge sort.
- fix incorrect result for distinct agg
- fix incorrect result for topn in local mode 

I'd like to fix the **FIXME** in a separate PR. We merge this PR if other fixes look good? 

cc @liurenjie1024 @fuyufjh

**Review by BugenZhao — APPROVED** [review]

**Comment by BugenZhao** [comment]
So this does not enable real parallel table scan according to discussions in #3583? I've removed "close" in PR body. 🤣

**Pull request merged**
